core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java

```diff
@@ -359,6 +360,55 @@ public boolean equalsKey(long address, WrappedBytes wrappedBytes) {
       return true;
    }
 
+   /**
+    * Returns whether entry is expired or not.
+    * @param address the address of the entry's key to check
+    * @return {@code true} if the entry is expired, {@code false} otherwise
+    */
+   @Override
+   public boolean isExpired(long address) {
+      // 16 bytes for eviction if needed (optional)
+      // 8 bytes for linked pointer
+      int offset = evictionEnabled ? 24 : 8;
+
+      byte metadataType = MEMORY.getByte(address, offset);
```

Comment:
In many cases we can just stop right after reading this. If the metadataType contains the IMMORTAL flag then we know it is not expired.